File: api/src/function/bibframe/Instance/instance.py
from rdflib import Graph, Namespace, URIRef, Literal
from rdflib.namespace import RDF, RDFS
from src.function.bibframe.Work.workAdmin import WorkAdmin
from src.function.bibframe.Work.title import Title
from src.function.bibframe.Instance.extent import Extent
from src.function.bibframe.Instance.provisionActivity import ProvisionActivity
from src.function.bibframe.Instance.responsibilityStatement import ResponsibilityStatement
from src.function.bibframe.Instance.seriesStatement import SeriesStatement
from src.function.bibframe.Instance.editionStatement import Edition
from src.function.bibframe.Instance.instanceOf import InstanceOf
from src.function.bibframe.Work.updateWork import UpdateWork

def BfInstance(request, instance_id):
     
    print("INSTSANCE: ", instance_id)
    instance_uri = URIRef(
        f"https://bibliokeia.com/resources/instance/{instance_id}")
    
    g = Graph(identifier=instance_uri)

    #Prefix
    g.bind('rdf', RDF)
    g.bind('rdfs', RDFS)
    BF = Namespace("http://id.loc.gov/ontologies/bibframe/")
    g.bind('bf', BF)
    BFLC = Namespace("http://id.loc.gov/ontologies/bflc/")
    g.bind('bflc', BFLC)
    MADSRDF = Namespace("http://www.loc.gov/mads/rdf/v1#")
    g.bind('madsrdf', MADSRDF)

    if request.subtitle: 
        label = Literal(f'{request.mainTitle}: {request.subtitle}')
    else:
        label = Literal(request.mainTitle)

    g.add((instance_uri, RDF.type, BF.Instance)) 
    g.add((instance_uri, RDF.type, BF.Print)) 
    g.add((instance_uri, RDFS.label, label)) 

    # AdminMetadata
    g = WorkAdmin(g, instance_uri, request.instanceOf, BF) 

    # Title
    g = Title(g, request, instance_uri, label, BF)

    # Edition
    g = Edition(g, request, instance_uri, BF)

    # Extent
    g = Extent(g, request, instance_uri, BF)

    # ProvisionActivity
    g = ProvisionActivity(g, request, instance_uri, BF)

    # responsibilityStatement
    g = ResponsibilityStatement(g, request, instance_uri, BF)

    # seriesStatement
    g = SeriesStatement(g, request, instance_uri, BF)

    g = InstanceOf(g, request, instance_uri, BF)
    UpdateWork(request.instanceOf, instance_id)

    return g

File: ui/src/services/acervo/instance.js
import fetch from "@rdfjs/fetch";
import cf from "clownface";
import namespace from "@rdfjs/namespace";

async function ParserInstance(id, setInstance) {
  const ns = {
    rdf: namespace("http://www.w3.org/1999/02/22-rdf-syntax-ns#"),
    rdfs: namespace("http://www.w3.org/2000/01/rdf-schema#"),
    bf: namespace("http://id.loc.gov/ontologies/bibframe/"),
    bflc: namespace("http://id.loc.gov/ontologies/bflc/"),
    owl: namespace("http://www.w3.org/2002/07/owl#"),
    skos: namespace("http://www.w3.org/2004/02/skos/core#"),
    dcterms: namespace("http://purl.org/dc/terms/"),
    cc: namespace("http://creativecommons.org/ns#"),
    foaf: namespace("http://xmlns.com/foaf/0.1/"),
    madsrdf: namespace("http://www.loc.gov/mads/rdf/v1#"),
  };

  let uri = `https://bibliokeia.com/resources/instance/${id}`;
  let graph = `http://localhost:3030/acervo?graph=${uri}`;

  let dataset = await fetch(graph).then((response) => response.dataset());

  let tbbt = cf({ dataset });
  let resources = tbbt.namedNode(uri);

  // Title
  const title = resources.out(ns.bf.title).out(ns.bf.mainTitle).value;

  // Type
  const types = resources.out(ns.rdf.type);
  const type = types.filter((type) => {
    return type.value.split("/")[5] != "Instance";
  });
  let typeLabel = type.value.split("/")[5];

  //Extent
  let extent = resources.out(ns.bf.extent).out(ns.rdfs.label).value;

  // ProvisionActivity
  let provisionActivity = resources.out(ns.bf.provisionActivity);
  let editora = provisionActivity.out(ns.bf.agent).out(ns.rdfs.label).value;
  let date = provisionActivity.out(ns.bf.date).value;
  let place = provisionActivity.out(ns.bf.place).out(ns.rdfs.label).value;

  // Items
  let hasItem = resources.out(ns.bf.hasItem).values

  let instanceOf = resources.out(ns.bf.instanceOf).value
  let work = instanceOf.split("/")[5]



  async function GetItems(uri) {
    let graphItems = `http://localhost:3030/acervo?graph=${uri}`;

    let dataset = await fetch(graphItems).then((response) =>
      response.dataset()
    );
    let tbbt = cf({ dataset });
    let item = tbbt.namedNode(uri);
    let label = item.out(ns.bf.shelfMark).out(ns.rdfs.label).value;
    let sublocation = item.out(ns.bf.sublocation).out(ns.rdfs.label).value;
    let register = item.out(ns.bf.adminMetadata).out(ns.bf.identifiedBy).out(ns.rdf.value).value
    
    return { uri: uri, label: label, sublocation: sublocation, register: register };
  }

  const promises = hasItem.map((uri) => GetItems(uri));
  let items = await Promise.all(promises).then((response) => {
    return response;
  });


  const instance = {
    title: title,
    typeLabel: typeLabel,
    extent: extent,
    editora: editora,
    place: place,
    date: date,
    instanceOf: work,
    items: items
  };

  setInstance(instance);
}

export default ParserInstance;
